src/main/java/bio/terra/service/job/JobService.java

```diff
@@ -54,32 +60,105 @@
     private final KubeService kubeService;
     private final JobShutdownState jobShutdownState;
+    private final Migrate migrate;
+
+    // TODO: this class reliably returns the Spring class loader whereas the JobService class does not.
+    //  It is a mystery worth solving at some point.
+    private final FireStoreDao fireStoreDao;
 
     @Autowired
     public JobService(IamService samService,
                       ApplicationConfiguration appConfig,
                       StairwayJdbcConfiguration stairwayJdbcConfiguration,
                       MigrateConfiguration migrateConfiguration,
+                      GoogleResourceConfiguration googleResourceConfiguration,
                       ApplicationContext applicationContext,
                       KubeService kubeService,
                       JobShutdownState jobShutdownState,
+                      Migrate migrate,
+                      FireStoreDao fireStoreDao,
                       ObjectMapper objectMapper) {
         this.samService = samService;
         this.appConfig = appConfig;
+        this.kubeService = kubeService;
         this.stairwayJdbcConfiguration = stairwayJdbcConfiguration;
         this.migrateConfiguration = migrateConfiguration;
-        this.kubeService = kubeService;
         this.jobShutdownState = jobShutdownState;
+        this.migrate = migrate;
+        this.fireStoreDao = fireStoreDao;
 
-        logger.info("Creating Stairway: maxStairwayThreads = " + appConfig.getMaxStairwayThreads());
+        String projectId = googleResourceConfiguration.getProjectId();
+        String stairwayClusterName = kubeService.getNamespace() + "-stairwaycluster";
+
+        logger.info("Creating Stairway: maxStairwayThreads: " + appConfig.getMaxStairwayThreads() +
+            " in project: " + projectId);
         ExceptionSerializer serializer = new StairwayExceptionSerializer(objectMapper);
         stairway = Stairway.newBuilder()
             .maxParallelFlights(appConfig.getMaxStairwayThreads())
             .exceptionSerializer(serializer)
             .applicationContext(applicationContext)
             .stairwayName(appConfig.getPodName())
             .stairwayHook(new StairwayLoggingHooks())
-        .build();
+            .stairwayClusterName(stairwayClusterName)
+            .projectId(projectId)
+            .enableWorkQueue(appConfig.isInKubernetes())
+            .classLoader(fireStoreDao.getClass().getClassLoader())
+            .build();
+    }
+
+    /**
+     * This method is called from StartupInitializer as part of the sequence of migrating databases
+     * and recovering any jobs; i.e., Stairway flights. It lives in this class so that JobService encapsulates
+     * all Stairway interaction.
+     */
+    public void initialize() {
+        try {
+            // Order is important here. We start the kubeListener so we see any pod deletes as we are starting up.
+            // Then we initialize Stairway and gets its list of recorded pods.
```

Comment:
Isn't this the opposite of the order below? (stairway.initialize then kubeservice.startpodlistener).

Reply:
I moved and re-wrote the comment. It was definitely confusing as written.